<a href="https://colab.research.google.com/github/Atsoutse1/git-jedha-ats/blob/main/02_Tidying_up_Part_1_solutions_Databricks.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Tidying up our data - Part 1

# Learning objectives

- Manipulate deeply nested json and transform it into structured data ready to be loaded onto a Data Warehousing system

## Loading our data from S3

Because we expect you to know how to setup things so that you can load data from S3, we're doing it for you now.  
Make sure you go through our code and check that you were actually using our best practices.

---

On a side note, it is **tedious** to have to do this in every notebook, isn't it?  
And putting everything into a single notebook isn't a proper solution either.

The solution would be to store this configuration into an external file, that's what we would recommend.  
_Sadly, DataBricks doesn't work with regular python files, only notebooks..._

In [ ]:
!pip install boto3

Requirement already satisfied: boto3 in /databricks/python3/lib/python3.7/site-packages (1.12.0)
Requirement already satisfied: botocore<1.16.0,>=1.15.0 in /databricks/python3/lib/python3.7/site-packages (from boto3) (1.15.0)
Requirement already satisfied: jmespath<1.0.0,>=0.7.1 in /databricks/python3/lib/python3.7/site-packages (from boto3) (0.10.0)
Requirement already satisfied: s3transfer<0.4.0,>=0.3.0 in /databricks/python3/lib/python3.7/site-packages (from boto3) (0.3.3)
Requirement already satisfied: docutils<0.16,>=0.10 in /databricks/python3/lib/python3.7/site-packages (from botocore<1.16.0,>=1.15.0->boto3) (0.15.2)
Requirement already satisfied: python-dateutil<3.0.0,>=2.1 in /databricks/python3/lib/python3.7/site-packages (from botocore<1.16.0,>=1.15.0->boto3) (2.8.1)
Requirement already satisfied: urllib3<1.26,>=1.20; python_version != "3.4" in /databricks/python3/lib/python3.7/site-packages (from botocore<1.16.0,>=1.15.0->boto3) (1.25.8)
Requirement already satisfied: six>=1.5 in /databricks/python3/lib/python3.7/site-packages (from python-dateutil<3.0.0,>=2.1->botocore<1.16.0,>=1.15.0->boto3) (1.14.0)
 WARNING: You are using pip version 20.0.2; however, version 20.3.3 is available.
You should consider upgrading via the '/databricks/python3/bin/python3.7 -m pip install --upgrade pip' command.

In [ ]:
import boto3

In [ ]:
# TODO: using your previously created constants:
#       - create a s3 resource
#       - and a s3 bucket
### BEGIN STRIP ###
ACCESS_KEY_ID = "AKIA3V3GLDX54DRFGYTB" # cle du compte student
SECRET_ACCESS_KEY = "xpGN4+hrbJTcyxjBGtiKEDpPo46g+wuTKzo6wDGe" # secret key du compte student

hadoop_conf = spark._jsc.hadoopConfiguration()
hadoop_conf.set("fs.s3a.access.key", ACCESS_KEY_ID)
hadoop_conf.set("fs.s3a.secret.key", SECRET_ACCESS_KEY)
hadoop_conf.set("fs.s3a.impl","org.apache.hadoop.fs.s3a.S3AFileSystem") 

BUCKET_NAME = "full-stack-bigdata-datasets"
PREFIX = "Big_Data/YOUTUBE"
SCHEME = "s3"

session = boto3.Session(
    region_name='eu-west-3',  # Datacenters located in Paris, FR
    aws_access_key_id=ACCESS_KEY_ID, 
    aws_secret_access_key=SECRET_ACCESS_KEY
)
s3 = session.resource('s3')

### END STRIP ###

In [ ]:
# We create a S3 resource and a Bucket from this same resource
bucket = s3.Bucket(BUCKET_NAME)

Last time, we worked with one file at a time, that was ok for quick analysis, but for processing, it would be nice to handle all files at once.  
We just gotta make sure their schemas is the same or we might run into surprises.  
Since all the APIs calls have been made using the same version and settings of the API, we can expect thing to be right, but **always keep an eye open**.

In [ ]:
# TODO: print out the list of files inside {BUCKET}/{PREFIX}
### BEGIN STRIP ###
all_keys = [obj.key for obj in bucket.objects.filter(Prefix=PREFIX)]
all_keys
### END STRIP ###

Out[20]: ['Big_Data/YOUTUBE/',
 'Big_Data/YOUTUBE/8cceb81c-534b-4354-8930-c7950ff5492c/_SUCCESS',
 'Big_Data/YOUTUBE/8cceb81c-534b-4354-8930-c7950ff5492c/_committed_5739506652000659913',
 'Big_Data/YOUTUBE/8cceb81c-534b-4354-8930-c7950ff5492c/_started_5739506652000659913',
 'Big_Data/YOUTUBE/8cceb81c-534b-4354-8930-c7950ff5492c/manifest.json',
 'Big_Data/YOUTUBE/8cceb81c-534b-4354-8930-c7950ff5492c/part-00000-tid-5739506652000659913-f0c4febb-409e-4b9c-b1ac-1c88875aaaf7-14-1-c000.csv',
 'Big_Data/YOUTUBE/interim/SS/df_songs_flattened.parquet/_SUCCESS',
 'Big_Data/YOUTUBE/interim/SS/df_songs_flattened.parquet/_committed_232945649319820793',
 'Big_Data/YOUTUBE/interim/SS/df_songs_flattened.parquet/_started_232945649319820793',
 'Big_Data/YOUTUBE/interim/SS/df_songs_flattened.parquet/part-00000-tid-232945649319820793-7dd8fd13-47cd-470e-8753-45c64e44c2b8-387-1-c000.snappy.parquet',
 'Big_Data/YOUTUBE/interim/SS/items_exploded.parquet/_committed_4074629136918237922',
 'Big_Data/YOUTUBE/interim/SS/items_exploded.parquet/_committed_5741717508296265154',
 'Big_Data/YOUTUBE/interim/SS/items_exploded.parquet/_committed_579680027546317307',
 'Big_Data/YOUTUBE/interim/SS/items_exploded.parquet/_started_579680027546317307',
 'Big_Data/YOUTUBE/interim/SS/items_exploded.parquet/part-00000-tid-579680027546317307-00c99957-2c62-4b2b-a54d-49ba06c8436c-404-1-c000.snappy.parquet',
 'Big_Data/YOUTUBE/interim/df_songs_flattened.parquet/_SUCCESS',
 'Big_Data/YOUTUBE/interim/df_songs_flattened.parquet/_committed_2377028696738576594',
 'Big_Data/YOUTUBE/interim/df_songs_flattened.parquet/_committed_998868613129258922',
 'Big_Data/YOUTUBE/interim/df_songs_flattened.parquet/_committed_vacuum5389043483423529704',
 'Big_Data/YOUTUBE/interim/df_songs_flattened.parquet/_started_998868613129258922',
 'Big_Data/YOUTUBE/interim/df_songs_flattened.parquet/part-00000-tid-998868613129258922-35122b34-9edf-4f73-96b8-36c5836889a9-9-1-c000.snappy.parquet',
 'Big_Data/YOUTUBE/items_clean.parquet/_committed_2102500016680538584',
 'Big_Data/YOUTUBE/items_clean.parquet/_committed_267942528307577806',
 'Big_Data/YOUTUBE/items_clean.parquet/_committed_3041171721563952764',
 'Big_Data/YOUTUBE/items_clean.parquet/_committed_3504085451804041009',
 'Big_Data/YOUTUBE/items_clean.parquet/_committed_3794523663304041093',
 'Big_Data/YOUTUBE/items_clean.parquet/_committed_4124778228644789872',
 'Big_Data/YOUTUBE/items_clean.parquet/_committed_4225566553644373009',
 'Big_Data/YOUTUBE/items_clean.parquet/_committed_4274277593579088726',
 'Big_Data/YOUTUBE/items_clean.parquet/_committed_4445695689079754518',
 'Big_Data/YOUTUBE/items_clean.parquet/_committed_4949729137628770263',
 'Big_Data/YOUTUBE/items_clean.parquet/_committed_5117470038809247762',
 'Big_Data/YOUTUBE/items_clean.parquet/_committed_5318146606096679769',
 'Big_Data/YOUTUBE/items_clean.parquet/_committed_5471106144365973191',
 'Big_Data/YOUTUBE/items_clean.parquet/_committed_5841726447239535840',
 'Big_Data/YOUTUBE/items_clean.parquet/_committed_7144498549587601004',
 'Big_Data/YOUTUBE/items_clean.parquet/_committed_7424108551056343149',
 'Big_Data/YOUTUBE/items_clean.parquet/_committed_7542173432757322126',
 'Big_Data/YOUTUBE/items_clean.parquet/_committed_7778300921098540745',
 'Big_Data/YOUTUBE/items_clean.parquet/_committed_7907349097119628941',
 'Big_Data/YOUTUBE/items_clean.parquet/_committed_7955925554486108274',
 'Big_Data/YOUTUBE/items_clean.parquet/_committed_8232636858920658403',
 'Big_Data/YOUTUBE/items_clean.parquet/_committed_8276126398753373234',
 'Big_Data/YOUTUBE/items_clean.parquet/_committed_8317291788803166763',
 'Big_Data/YOUTUBE/items_clean.parquet/_committed_8449078668241963530',
 'Big_Data/YOUTUBE/items_clean.parquet/_committed_8677531215091547790',
 'Big_Data/YOUTUBE/items_clean.parquet/_committed_8699957198574130244',
 'Big_Data/YOUTUBE/items_clean.parquet/_committed_9032510107735312434',
 'Big_Data/YOUTUBE/items_clean.parquet/_committed_9077756668064582686',
 'Big_Data/YOUTUBE/items_clean.parque

As you can see in the [documentation](https://spark.apache.org/docs/2.1.0/api/python/pyspark.sql.html#pyspark.sql.DataFrameReader.load) for the `.load(...)` method, it accepts as `path` argument, either a string or a list of string.

In [ ]:
# TODO: write the filepath using SCHEME, BUCKET_NAME, PREFIX: `filepath`
### BEGIN STRIP ###
file_paths = "s3://full-stack-bigdata-datasets/Big_Data/YOUTUBE/songs.json"
file_paths
### END STRIP ###

Out[21]: 's3://full-stack-bigdata-datasets/Big_Data/YOUTUBE/songs.json'

In [ ]:
# TODO: load all the files into a DataFrame: `df`
### BEGIN STRIP ###
df = spark.read.format('json').load(file_paths, multiline=True)
### END STRIP ###

In [ ]:
# TODO: count the number of rows in your DataFrame, it should be 100
### BEGIN STRIP ###
df.count()
### END STRIP ###

Out[23]: 100

Looking great! But there's more, what they're not telling you in the documentation is that you can use wildcards in your paths:
- `*`: replaces any string
- `?`: replaces a single letter

This pattern is called globbing, you can learn about it on [Wikipedia](https://en.wikipedia.org/wiki/Glob_(programming).

We'll just check we have the same number of rows with this method

In [ ]:
# TODO: count the number of rows in the DataFrame
### BEGIN STRIP ###
df.count()
### END STRIP ###

Out[24]: 100

Once again, we should have 100 rows. **If that's not the case, you have an issue, go back and fix it.**

## Tidying up

---

We have multiple issues with our data.  **It does not look like "tidy data" at all.**  
First, we have rows within rows...
And second, most of the data resides in deeply nested structure within the column items...

We will fix the former, then handle the latter in the next notebook.

### 1. Fixing the rows
Ever heard about `EXPLODE` in SQL?

Luckily for us, they're an equivalent in PySpark: `.explode(...)`, here's the link to the [documentation](https://spark.apache.org/docs/2.1.0/api/python/pyspark.sql.html#pyspark.sql.functions.explode).  
What `.explode(...)` does, it "Returns a new row for each element in the given array or map."

If you remember properly, that's exactly the kind of structures we have in the schema of our DataFrame for the `items` column.

In [ ]:
# TODO: print out the schema of `df`
### BEGIN STRIP ###
df.printSchema()
### END STRIP ###

root
-- etag: string (nullable = true)
-- items: array (nullable = true)
 |-- element: struct (containsNull = true)
 | |-- contentDetails: struct (nullable = true)
 | | |-- caption: string (nullable = true)
 | | |-- contentRating: struct (nullable = true)
 | | | |-- ytRating: string (nullable = true)
 | | |-- definition: string (nullable = true)
 | | |-- dimension: string (nullable = true)
 | | |-- duration: string (nullable = true)
 | | |-- licensedContent: boolean (nullable = true)
 | | |-- projection: string (nullable = true)
 | | |-- regionRestriction: struct (nullable = true)
 | | | |-- allowed: array (nullable = true)
 | | | | |-- element: string (containsNull = true)
 | | | |-- blocked: array (nullable = true)
 | | | | |-- element: string (containsNull = true)
 | |-- etag: string (nullable = true)
 | |-- id: string (nullable = true)
 | |-- kind: string (nullable = true)
 | |-- snippet: struct (nullable = true)
 | | |-- categoryId: string (nullable = true)
 | | |-- channelId: string (nullable = true)
 | | |-- channelTitle: string (nullable = true)
 | | |-- defaultAudioLanguage: string (nullable = true)
 | | |-- defaultLanguage: string (nullable = true)
 | | |-- description: string (nullable = true)
 | | |-- liveBroadcastContent: string (nullable = true)
 | | |-- localized: struct (nullable = true)
 | | | |-- description: string (nullable = true)
 | | | |-- title: string (nullable = true)
 | | |-- publishedAt: string (nullable = true)
 | | |-- tags: array (nullable = true)
 | | | |-- element: string (containsNull = true)
 | | |-- thumbnails: struct (nullable = true)
 | | | |-- default: struct (nullable = true)
 | | | | |-- height: long (nullable = true)
 | | | | |-- url: string (nullable = true)
 | | | | |-- width: long (nullable = true)
 | | | |-- high: struct (nullable = true)
 | | | | |-- height: long (nullable = true)
 | | | | |-- url: string (nullable = true)
 | | | | |-- width: long (nullable = true)
 | | | |-- maxres: struct (nullable = true)
 | | | | |-- height: long (nullable = true)
 | | | | |-- url: string (nullable = true)
 | | | | |-- width: long (nullable = true)
 | | | |-- medium: struct (nullable = true)
 | | | | |-- height: long (nullable = true)
 | | | | |-- url: string (nullable = true)
 | | | | |-- width: long (nullable = true)
 | | | |-- standard: struct (nullable = true)
 | | | | |-- height: long (nullable = true)
 | | | | |-- url: string (nullable = true)
 | | | | |-- width: long (nullable = true)
 | | |-- title: string (nullable = true)
 | |-- statistics: struct (nullable = true)
 | | |-- commentCount: string (nullable = true)
 | | |-- dislikeCount: string (nullable = true)
 | | |-- favoriteCount: string (nullable = true)
 | | |-- likeCount: string (nullable = true)
 | | |-- viewCount: string (nullable = true)
 | |-- status: struct (nullable = true)
 | | |-- embeddable: boolean (nullable = true)
 | | |-- license: string (nullable = true)
 | | |-- madeForKids: boolean (nullable = true)
 | | |-- privacyStatus: string (nullable = true)
 | | |-- publicStatsViewable: boolean (nullable = true)
 | | |-- uploadStatus: string (nullable = true)
 | |-- topicDetails: struct (nullable = true)
 | | |-- relevantTopicIds: array (nullable = true)
 | | | |-- element: string (containsNull = true)
 | | |-- topicCategories: array (nullable = true)
 | | | |-- element: string (containsNull = true)
-- kind: string (nullable = true)
-- pageInfo: struct (nullable = true)
 |-- resultsPerPage: long (nullable = true)
 |-- totalResults: long (nullable = true)

In [ ]:
# TODO: import the PySpark SQL functions following usual convention
### BEGIN STRIP ###
from pyspark.sql import functions as F
### END STRIP ###

In [ ]:
# TODO: use `.explode(...)` on the `items` column and count the number of results
### BEGIN STRIP ###
df.select(F.explode('items')).count()
### END STRIP ###

Out[27]: 3907

If you got 3907 rows, you've made it, congrats! :)  
We will use this as our new working DataFrame:
- just do the same thing, but this time save into a variable named `items_df`
- don't forget to give a proper alias to your newly compute column: `items`
- at the end, as a sanity check, make sure we have the right amount of columns in our new DataFrame

In [ ]:
# TODO: follow previous instructions
### BEGIN STRIP ###
items_df = df.select(F.explode('items').alias('items'))
items_df.count()
### END STRIP ###

Out[28]: 3907

We're making progress, we now have one row per result (e.g. song)!

But each song is a deeply nested structure... We will take care of this in the following notebook.

In [ ]:
# TODO: save the DataFrame as a parquet
### BEGIN STRIP ###
# items_exploded_path = "s3://full-stack-bigdata-datasets/Big_Data/YOUTUBE/items_exploded.parquet"
# items_df.write.parquet(items_exploded_path, mode='overwrite')
### END STRIP ###